Fix issue #832 - removing federation upstream

fiksn opened the pull request:
<!--
Thank you for contributing to this project!

- This project has a Contributor Code of Conduct: https://voxpupuli.org/coc/
- Please check that here is no existing issue or PR that addresses your problem.
- Our vulnerabilities reporting process is at https://voxpupuli.org/security/

-->
#### Pull Request (PR) description
<!--
Replace this comment with a description of your pull request.
-->
I just fixed the bug, there should be some better unit test as well.

#### This Pull Request (PR) fixes the following issues
<!--
Replace this comment with the list of issues or n/a.
Use format:
Fixes #123
Fixes #124
-->
Fixes #832

bastelfreak commented:
@fiksn thanks for fixing this, can you somehow add a test to ensure this works now as expected?

fiksn commented:
@bastelfreak: Yeah, adding tests would be my wish too, but unfortunately I am not that fluent with Ruby (yet). For now the test on my side is referencing my repo with the change in Puppetfile.

fiksn commented:
I've added some unit tests and ran the stuff with `docker build .` It fails with the old version of code and passes with the new. So I suppose it should be good enough 🙂

wyardley commented:
First glance, LGTM, but haven’t been doing much Puppet or Ruby development recently, so probably good to get some eyes on it, especially in terms of whether the tests test the right things.

wyardley commented on Dockerfile:
```diff
@@ -1,6 +1,7 @@
 FROM ruby:2.5.3
 
 WORKDIR /opt/puppet
+ENV LC_ALL=C.UTF-8
```
Is this needed or just a byproduct of something you added for testing locally? Not sure if this file comes from modulesync

fiksn replied:
Apparently the default locale is set to "POSIX" and when I run the test through docker container rake fails with "ArgumentError: invalid byte sequence in US-ASCII". I agree that this might be a bit offtopic for this issue, but fact is that without that line `docker build .` fails. Of course if I installed and ran everything locally stuff still passes.

Not sure what should be done here. Perhaps I can squash my two commits that deal with this bug and then create a separate commit / pull request for the issue that test failure when using docker. Perhaps there is even some nicer way (like that it is expected to pass some -e LC_ALL.. flag, but didn't find any documentation about it).

ekohl commented on spec/unit/puppet/provider/rabbitmq_parameter/rabbitmqctl_federation_spec.rb:
```diff
@@ -0,0 +1,74 @@
+    end
+
+    it 'matches' do
+      provider_class.expects(:rabbitmqctl_list).with('vhosts').returns <<-EOT
```
In our modules we've been moving away from setting `expects` before the test but rather use `allow()`, then the test and then `expect().to have_received()`.
https://relishapp.com/rspec/rspec-mocks/docs/basics/spies#spy-on-a-method-on-a-partial-double

wyardley replied:
Makes sense; given that the submitter said they’re not very experienced with tests, and that this matches the style of the rest of the tests, I’m inclined to say let’s accept this, and then someone can refactor the tests to the newer style later if it makes sense?

ekohl replied:
I'm fine with that.

fiksn commented:
I am more than happy to contribute. Sorry for not being responsive lately.

Removed the part from Dockerfile now and squashed the two commits (fix and unit tests). Passing environment variables won't work as easy as I thought since this is the "build" and not "run" phase (you can just add --build-arg). Do you believe it makes sense to open another issue for that one-liner?

wyardley commented:
@fiksn so, I'm not doing too much work on Vox modules anymore, but the `Dockerfile` comes from https://github.com/voxpupuli/modulesync_config, so we can't make modifications directly to it in this module vs. upstream.

We have seen a few issues with locale settings like `LANG` or `LC_ALL` in this module, and there are some existing workarounds (e.g., 6c7a1467370ac926b48d6aa3c7aed4a78301b175) related to the CLI tools needing those to be set. These issues would only come up on systems where it wasn't set system-wide, so this may be a similar kind of issue.

It may be that more of those need to be made, vs. adjusting the environment in the test env, as this could actually be reflecting a real problem in cases where that env isn't set at system level. As the tests work in CI, I think we can assume the module is working, for the most part.
